### DISCUSSION TITLE
Passing ref.read or data itself?

### DISCUSSION THREAD

**Discussion by peterhijma** [rrousselGit/riverpod#819] [Q&A] [answered]
Right now I have these providers, where `userSessionProvider` is initialised on the start of the app (and therefore not `null`):

```dart
final userSessionProvider = StateProvider<UserSession?>((ref) => null);

final httpClientProvider = Provider<HttpClient>(
  (ref) => HttpClient(
    reader: ref.read,
  ),
);
```

Inside the `HttpClient` I use a `reader` to get the authentication token from the `userSessionProvider` (`UserSession.token`)

or ...  is it better to pass the `UserSession` like so:

```dart
final userSessionProvider = StateProvider<UserSession?>((ref) => null);

final httpClientProvider = Provider<HttpClient>(
  (ref) => HttpClient(
    userSession: ref.watch(userSessionProvider).state,
  ),
);
```

Or should the `HttpClient` know nothing about `UserSession` and should I pass the token directly...

```dart
final httpClientProvider = Provider<HttpClient>(
  (ref) => HttpClient(
    token: ref.watch(userSessionProvider).state!.token,
  ),
);
```

What's best practice here?

**Comment by rrousselGit** [comment, marked as answer]
They are all valid. There's no real answer to this question

**Reply by peterhijma** [reply]
Alright. Thanks! I thought maybe some would be easier when writing tests for example. But I guess it’s not a riverpod question, but more of general programming question, so it probably doesn’t belong here 😋
